File: Helpers/MunkiApps.cs
using System.Collections;
using PropertyList;
using SupportCompanion.Services;

namespace SupportCompanion.Helpers;

public class MunkiApps
{
    private readonly LoggerService _logger;
    private readonly string _managedInstallsReportPlist = "/Library/Managed Installs/ManagedInstallReport.plist";
    private readonly string _selfServeManifest = "/Library/Managed Installs/manifests/SelfServeManifest";

    // Default constructor
    public MunkiApps() : this(new LoggerService())
    {
    }

    // Constructor with LoggerService parameter
    private MunkiApps(LoggerService logger)
    {
        _logger = logger;
    }

    private async Task<Stream?> ReadFileWithRetry(string filePath, int maxRetries = 3,
        int delayMilliseconds = 20000)
    {
        for (var i = 0; i < maxRetries; i++)
            try
            {
                return File.Open(filePath, FileMode.Open, FileAccess.Read, FileShare.ReadWrite);
            }
            catch (IOException)
            {
                // Wait before retrying
                _logger.Log(
                    "ReadFileWithRetry",
                    $"File {filePath} is not available. Retrying in {delayMilliseconds} milliseconds.",
                    1);
                await Task.Delay(delayMilliseconds);
            }

        // If we've tried to open the file maxRetries times and it's still not available, log an error and return null
        _logger.Log("ReadFileWithRetry",
            $"File {filePath} could not be opened after {maxRetries} attempts.", 2);
        return null;
    }

    private async Task LookForFileWithRetry(string filePath, int maxRetries = 3, int delayMilliseconds = 20000)
    {
        for (var i = 0; i < maxRetries; i++)
            if (File.Exists(filePath))
            {
                return;
            }
            else
            {
                // Wait before retrying
                _logger.Log(
                    "LookForFileWithRetry",
                    $"File {filePath} is not available. Retrying in {delayMilliseconds} milliseconds.",
                    1);
                await Task.Delay(delayMilliseconds);
            }

        // If we've tried to open the file maxRetries times and it's still not available, log an error and return null
        _logger.Log("LookForFileWithRetry",
            $"File {filePath} could not be found after {maxRetries} attempts.", 2);
    }

    public async Task<int> GetPendingUpdates()
    {
        await LookForFileWithRetry(_managedInstallsReportPlist);
        await using var reader = await ReadFileWithRetry(_managedInstallsReportPlist);
        if (reader == null)
        {
            _logger.Log("MunkiApps:GetPendingUpdates", "Reader is null", 1);
            return 0;
        }

        var plistReader = new PlistReader();
        var plist = plistReader.Read(reader);
        if (plist.TryGetValue("ItemsToInstall", out var pendingApps))
            return ((IList)pendingApps).Count;

        _logger.Log("MunkiApps:GetPendingUpdates", "ItemsToInstall key not found", 1);
        return 0;
    }

    public async Task<int> GetInstalledAppsCount()
    {
        await LookForFileWithRetry(_managedInstallsReportPlist);
        await using var reader = await ReadFileWithRetry(_managedInstallsReportPlist);
        if (reader == null)
        {
            _logger.Log("MunkiApps:GetInstalledAppCount", "Reader is null", 1);
            return 0;
        }

        var plistReader = new PlistReader();
        var plist = plistReader.Read(reader);
        if (plist.TryGetValue("InstalledItems", out var installedApps))
            return ((IList)installedApps).Count;

        _logger.Log("MunkiApps:GetInstalledAppCount", "InstalledItems key not found", 1);
        return 0;
    }

    public async Task<IList> GetPendingUpdatesList()
    {
        await LookForFileWithRetry(_managedInstallsReportPlist);
        await using var reader = await ReadFileWithRetry(_managedInstallsReportPlist);
        if (reader == null)
        {
            _logger.Log("MunkiApps:GetPendingUpdatesList", "Reader is null", 1);
            return new List<string>();
        }

        var plistReader = new PlistReader();
        var plist = plistReader.Read(reader);
        if (plist.TryGetValue("ItemsToInstall", out var pendingApps))
            return (IList)pendingApps;

        _logger.Log("MunkiApps:GetPendingUpdatesList", "ItemsToInstall key not found", 1);
        return new List<string>();
    }

    public async Task<IList> GetInstalledAppsList()
    {
        await LookForFileWithRetry(_managedInstallsReportPlist);
        await using var reader = await ReadFileWithRetry(_managedInstallsReportPlist);
        if (reader == null)
        {
            _logger.Log("MunkiApps:GetInstalledAppsList", "Reader is null", 1);
            return new List<string>();
        }

        var plistReader = new PlistReader();
        var plist = plistReader.Read(reader);
        if (plist.TryGetValue("ManagedInstalls", out var installedApps))
            return (IList)installedApps;

        _logger.Log("MunkiApps:GetInstalledAppsList", "ManagedInstalls key not found", 1);
        return new List<string>();
    }

    public async Task<IList> GetSelfServeAppsList()
    {
        await LookForFileWithRetry(_selfServeManifest);
        await using var reader = await ReadFileWithRetry(_selfServeManifest);
        if (reader == null)
        {
            _logger.Log("MunkiApps:GetSelfServeAppsList", "Reader is null", 1);
            return new List<string>();
        }

        var plistReader = new PlistReader();
        var plist = plistReader.Read(reader);
        if (plist.TryGetValue("managed_installs", out var selfServeApps))
            return (IList)selfServeApps;

        _logger.Log("MunkiApps:GetSelfServeAppsList", "managed_installs key not found", 1);
        return new List<string>();
    }
}

File: SupportCompanion/Helpers/MunkiApps.swift
//
//  MunkiApps.swift
//  SupportCompanion
//
//  Created by Tobias Almén on 2024-11-20.
//

import Foundation

class MunkiApps {
    private let logger = Logger.shared // Replace with your logging service
    private let managedInstallsReportPlist = "/Library/Managed Installs/ManagedInstallReport.plist"
    private let selfServeManifest = "/Library/Managed Installs/manifests/SelfServeManifest"
    
    /// Attempts to read a file with retries
    private func readFileWithRetry(filePath: String, maxRetries: Int = 3, delayMilliseconds: Int = 20000) async throws -> Data? {
        for attempt in 1...maxRetries {
            do {
                return try Data(contentsOf: URL(fileURLWithPath: filePath))
            } catch {
                if attempt < maxRetries {
                    logger.logDebug("File \(filePath) not available. Retrying in \(delayMilliseconds)ms.")
                    try await Task.sleep(nanoseconds: UInt64(delayMilliseconds) * 1_000_000)
                } else {
                    logger.logError("File \(filePath) could not be opened after \(maxRetries) attempts.")
                    throw error
                }
            }
        }
        return nil
    }
    
    /// Ensures a file exists with retries
    private func lookForFileWithRetry(filePath: String, maxRetries: Int = 3, delayMilliseconds: Int = 2000) async throws {
        for attempt in 1...maxRetries {
            if FileManager.default.fileExists(atPath: filePath) {
                return
            }
            if attempt < maxRetries {
                logger.logDebug("File \(filePath) not found. Retrying in \(delayMilliseconds)ms.")
                try await Task.sleep(nanoseconds: UInt64(delayMilliseconds) * 1_000_000)
            } else {
                logger.logError("File \(filePath) could not be found after \(maxRetries) attempts.")
                throw NSError(domain: "MunkiApps", code: 1, userInfo: [NSLocalizedDescriptionKey: "File not found"])
            }
        }
    }
    
    /// Reads a plist and parses it into a dictionary
    private func readPlist(filePath: String) async throws -> [String: Any] {
        try await lookForFileWithRetry(filePath: filePath)
        guard let data = try await readFileWithRetry(filePath: filePath) else {
            logger.logError("Failed to read file: \(filePath)")
            throw NSError(domain: "MunkiApps", code: 2, userInfo: [NSLocalizedDescriptionKey: "Failed to read file"])
        }
        let plist = try PropertyListSerialization.propertyList(from: data, options: [], format: nil)
        guard let dictionary = plist as? [String: Any] else {
            logger.logError("Invalid plist format for file: \(filePath)")
            throw NSError(domain: "MunkiApps", code: 3, userInfo: [NSLocalizedDescriptionKey: "Invalid plist format"])
        }
        return dictionary
    }

    /// Returns the count of pending updates
    func getPendingUpdates() async -> Int {
        do {
            let plist = try await readPlist(filePath: managedInstallsReportPlist)
            if let itemsToInstall = plist["ItemsToInstall"] as? [Any] {
                return itemsToInstall.count
            } else {
                logger.logDebug("ItemsToInstall key not found")
                return 0
            }
        } catch {
            logger.logError("Error fetching pending updates: \(error)")
            return 0
        }
    }
    
    /// Returns the count of installed apps
    func getInstalledAppsCount() async -> Int {
        do {
            let plist = try await readPlist(filePath: managedInstallsReportPlist)
            if let installedItems = plist["InstalledItems"] as? [Any] {
                return installedItems.count
            } else {
                logger.logDebug("InstalledItems key not found")
                return 0
            }
        } catch {
            logger.logError("Error fetching installed apps count: \(error)")
            return 0
        }
    }
    
    /// Returns a list of pending updates
    func getPendingUpdatesList() async -> [PendingMunkiUpdate] {
        do {
            let plist = try await readPlist(filePath: managedInstallsReportPlist)
            if let itemsToInstall = plist["ItemsToInstall"] as? [[String: Any]] {
                return itemsToInstall.compactMap { appDict in
                    guard let name = appDict["display_name"] as? String,
                          let version = appDict["version_to_install"] as? String else {
                        return nil
                    }
                    return PendingMunkiUpdate(name: name, version: version)
                }
            } else {
                logger.logDebug("ItemsToInstall key not found")
                return []
            }
        } catch {
            logger.logError("Error fetching pending updates list: \(error)")
            return []
        }
    }
    
    /// Returns a list of installed apps
    func getInstalledAppsList() async -> [[String: Any]] {
        do {
            let plist = try await readPlist(filePath: managedInstallsReportPlist)
            if let managedInstalls = plist["ManagedInstalls"] as? [[String: Any]] {
                return managedInstalls
            } else {
                logger.logDebug("ManagedInstalls key not found")
                return []
            }
        } catch {
            logger.logError("Error fetching installed apps list: \(error)")
            return []
        }
    }
    
    /// Returns a list of self-serve apps
    func getSelfServeAppsList() async -> [String] {
        do {
            let plist = try await readPlist(filePath: selfServeManifest)
            if let managedInstalls = plist["managed_installs"] as? [String] {
                return managedInstalls
            } else {
                logger.logDebug("managed_installs key not found")
                return []
            }
        } catch {
            logger.logError("Error fetching self-serve apps list: \(error)")
            return []
        }
    }
}
